ntp-daemon/src/server.rs

```diff
@@ -160,3 +154,86 @@ fn accept_packet(
         }
     }
 }
+
+#[derive(Debug)]
+struct TimestampedCache<T> {
+    elements: Vec<Option<(T, std::time::Instant)>>,
+}
+
+impl<T: std::hash::Hash + Eq> TimestampedCache<T> {
+    fn new(length: usize) -> Self {
+        Self {
+            // looks a bit odd, but prevents a `Clone` constraint
+            elements: std::iter::repeat_with(|| None).take(length).collect(),
+        }
+    }
+
+    fn index(&self, item: &T) -> usize {
+        use std::hash::Hasher;
+
+        let mut hasher = std::collections::hash_map::DefaultHasher::default();
+
+        item.hash(&mut hasher);
+
+        hasher.finish() as usize % self.elements.len()
+    }
+
+    fn insert(&mut self, item: T, timestamp: std::time::Instant) {
+        let index = self.index(&item);
+        self.elements[index] = Some((item, timestamp));
+    }
+
+    fn get(&self, item: &T) -> Option<std::time::Instant> {
+        let (existing, timestamp) = self.elements[self.index(item)].as_ref()?;
+
+        if existing == item {
+            Some(*timestamp)
+        } else {
+            None
+        }
+    }
+
+    fn is_allowed(
+        &mut self,
+        item: T,
+        timestamp: std::time::Instant,
+        cutoff: std::time::Duration,
+    ) -> bool {
+        match self.get(&item) {
+            None => {
+                self.insert(item, timestamp);
```

Comment:
Deduplicate the `insert()` call?

Reply:
that didn't quite work (because of T not being clone) but I did some refactoring